ca/crl.go

```diff
@@ -0,0 +1,164 @@
+	for _, issuer := range issuers {
+		issuersByNameID[issuer.Cert.NameID()] = issuer
+	}
+
+	if lifetime == 0 {
+		logger.Warningf("got zero for crl lifetime; setting to default 9 days")
+		lifetime = 9 * 24 * time.Hour
+	} else if lifetime >= 10*24*time.Hour {
+		return nil, fmt.Errorf("crl lifetime cannot be more than 10 days, got: %s", lifetime)
+	} else if lifetime <= 0*time.Hour {
+		return nil, fmt.Errorf("crl lifetime must be positive, got: %s", lifetime)
+	}
+
+	return &crlImpl{
+		issuers:  issuersByNameID,
+		lifetime: lifetime,
+		log:      logger,
+	}, nil
+}
+
+func (ci *crlImpl) GenerateCRL(stream capb.CRLGenerator_GenerateCRLServer) error {
+	rcs := make([]pkix.RevokedCertificate, 0)
+	var number *big.Int
+	var thisUpdate time.Time
+	var issuer *issuance.Issuer
+
+	got_metadata := false
+	for {
+		in, err := stream.Recv()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+
+		switch payload := in.Payload.(type) {
+		case *capb.GenerateCRLRequest_Metadata:
+			if got_metadata {
+				return errors.New("got more than one metadata message")
+			}
+			got_metadata = true
+
+			if payload.Metadata.IssuerNameID == 0 || payload.Metadata.ThisUpdate == 0 {
+				return errors.New("got incomplete metadata message")
+			}
+
+			number = big.NewInt(payload.Metadata.ThisUpdate)
+			thisUpdate = time.Unix(0, payload.Metadata.ThisUpdate)
+
+			var ok bool
+			issuer, ok = ci.issuers[issuance.IssuerNameID(payload.Metadata.IssuerNameID)]
+			if !ok {
+				return fmt.Errorf("got unrecognized IssuerNameID: %d", payload.Metadata.IssuerNameID)
+			}
+
+		case *capb.GenerateCRLRequest_Entry:
+			serial, err := core.StringToSerial(payload.Entry.Serial)
+			if err != nil {
+				return err
+			}
+
+			if payload.Entry.RevokedAt == 0 {
+				return errors.New("got empty or zero revocation timestamp")
+			}
+			revokedAt := time.Unix(0, payload.Entry.RevokedAt)
+
+			var extensions []pkix.Extension
+			if payload.Entry.Reason != 0 {
+				reasonBytes, err := asn1.Marshal(asn1.Enumerated(payload.Entry.Reason))
+				if err != nil {
+					return err
+				}
+
+				extensions = []pkix.Extension{
+					{
+						Id:    asn1.ObjectIdentifier{2, 5, 29, 21},
+						Value: reasonBytes,
+					},
+				}
+			}
+
+			rcs = append(rcs, pkix.RevokedCertificate{
+				SerialNumber:   serial,
+				RevocationTime: revokedAt,
+				Extensions:     extensions,
+			})
+
+			if len(rcs)%1000 == 0 {
+				ci.log.Debugf("Read %d crlEntries from input stream", len(rcs))
+			}
+
+		default:
+			return errors.New("got empty or malformed message in input stream")
+		}
+	}
+
+	if !got_metadata {
+		return errors.New("no crl metadata received")
+	}
+
+	template := x509.RevocationList{
+		RevokedCertificates: rcs,
+		Number:              number,
+		ThisUpdate:          thisUpdate,
+		NextUpdate:          thisUpdate.Add(-time.Second).Add(ci.lifetime),
+	}
+
+	crlBytes, err := x509.CreateRevocationList(
```

Comment:
This needs an audit log.

For OCSP, we log every certificate we generate OCSP for, plus its status (we should add reasonCode). We chunk these up into about 38 per log line, IIRC.

For CRLs, I think we should do something similar, though we can omit the certificates that are not listed in the CRL. So, once we have the list of RevokedCertificates, we should log N lines, containing M certificates per line, with their status and reasonCode. We should then log the final revocation list creation event, with the important parameters (ThisUpdate, NextUpdate, Number, length of RevokedCertificates, maybe summary stats of reasonCode?), a hash of the final CRL, and its size. It would be nice to be able to log the hash of the tbsCertList but I don't think Go's current APIs give us that.

Reply:
Let's consider this good for now, and file a followup issue to plumb through the reason code once #6198 is done.